src/xrpld/app/tx/detail/PermissionedDomainSet.cpp

```diff
@@ -0,0 +1,159 @@
+
+NotTEC
+PermissionedDomainSet::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featurePermissionedDomains))
+        return temDISABLED;
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto const credentials = ctx.tx.getFieldArray(sfAcceptedCredentials);
+    if (credentials.empty() || credentials.size() > PD_ARRAY_MAX)
+        return temMALFORMED;
+    for (auto const& credential : credentials)
+    {
+        if (!credential.isFieldPresent(sfIssuer) ||
+            !credential.isFieldPresent(sfCredentialType))
+        {
+            return temMALFORMED;
+        }
+        auto sz = credential.getFieldVL(sfCredentialType).size();
+        if (!sz || sz > 64)
+            return temMALFORMED;
+    }
+
+    auto const domain = ctx.tx.at(~sfDomainID);
+    if (domain && *domain == beast::zero)
+        return temMALFORMED;
+
+    return preflight2(ctx);
+}
+
+TER
+PermissionedDomainSet::preclaim(PreclaimContext const& ctx)
+{
+    if (!ctx.view.read(keylet::account(ctx.tx.getAccountID(sfAccount))))
+        return tefINTERNAL;
+
+    auto const credentials = ctx.tx.getFieldArray(sfAcceptedCredentials);
+    for (auto const& credential : credentials)
+    {
+        if (!ctx.view.read(keylet::account(credential.getAccountID(sfIssuer))))
+            return temBAD_ISSUER;
+    }
+
+    if (!ctx.tx.isFieldPresent(sfDomainID))
+        return tesSUCCESS;
+    auto const domain = ctx.tx.getFieldH256(sfDomainID);
+    auto const sleDomain = ctx.view.read(keylet::permissionedDomain(domain));
+    if (!sleDomain)
+        return tecNO_ENTRY;
+    auto const owner = sleDomain->getAccountID(sfOwner);
+    auto account = ctx.tx.getAccountID(sfAccount);
+    if (owner != account)
+        return temINVALID_ACCOUNT_ID;
+
+    return tesSUCCESS;
+}
+
+/** Attempt to create the Permissioned Domain. */
+TER
+PermissionedDomainSet::doApply()
+{
+    auto const ownerSle = view().peek(keylet::account(account_));
+
+    // The purpose of this lambda is to modify the SLE for either creating a
+    // new or updating an existing object, to reduce code repetition.
+    // All checks have already been done. Just update credentials. Same logic
+    // for either new domain or updating existing.
+    // Silently remove duplicates.
+    auto updateSle = [this](std::shared_ptr<STLedgerEntry> const& sle) {
+        auto const& credentials = ctx_.tx.getFieldArray(sfAcceptedCredentials);
+
+        std::map<
+            std::pair<AccountID, Blob>,
+            std::reference_wrapper<STObject const>>
+            credMap;
+        for (auto const& c : credentials)
+            credMap.insert(
+                {{c.getAccountID(sfIssuer), c.getFieldVL(sfCredentialType)},
+                 c});
+
+        STArray credSorted;
+        credSorted.reserve(credMap.size());
+        for (auto const& [k, v] : credMap)
+            credSorted.push_back(v);
+        sle->setFieldArray(sfAcceptedCredentials, credSorted);
```

Comment:
I would prefer if this code used the credentials helper functions, so it's easier to read/audit.

Reply:
Behavior aligned with  XLS-70